File: solutions/Problem-4.java
// Sorting //

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> res = new HashMap<>();
        for (String s : strs) {
            char[] charArray = s.toCharArray();
            Arrays.sort(charArray);
            String sortedS = new String(charArray);
            res.putIfAbsent(sortedS, new ArrayList<>());
            res.get(sortedS).add(s);
        }
        return new ArrayList<>(res.values());
    }
}

// Hash Table //

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> res = new HashMap<>();
        //creates a new HashMap to evaluate through
        for (String s : strs) {
            int[] count = new int[26];
            //creates alphabet
            for (char c : s.toCharArray()) {
                count[c - 'a']++;
                //subtracts each ascii letter in the alphabet to count amount of letters
            }
            String key = Arrays.toString(count);
            //creates an easy way to grab our solutions
            res.putIfAbsent(key, new ArrayList<>());
            //inserts key into our map
            res.get(key).add(s);
            //inserts solutions into res
        }
        return new ArrayList<>(res.values());
        //returns our anagram solutions
    }
}

File: solutions/Problem-4.js
// Sorting //

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const res = {};
    for (let s of strs) {
        const sortedS = s.split("").sort().join("");
        if (!res[sortedS]) {
            res[sortedS] = []
        }

        res[sortedS].push(s);
    }
    
    return Object.values(res);
};

// Hash Table //

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const res = {};
    for (let s of strs) {
        const count = new Array(26).fill(0);
        //fills array with zeros equal to the amount of the alphabet
        for (let c of s) {
            //for every character in each string
            count[c.charCodeAt(0) - "a".charCodeAt(0)] += 1;
            //subtract teh ascii value of a from the char at 0, then increment
            //to go up the alphabet to count each character
        } 

        const key = count.join(",");
        //joins our anagrams together
        if(!res[key]) {
            res[key] = [];
            //initates an empty array when a res does not exist
        }

        res[key].push(s);
        //pushes our result into the res
    }
    
    return Object.values(res);
};